pkg/util/test/leak.go

```diff
@@ -28,6 +28,9 @@ func goLeakOptions() []goleak.Option {
 		// it gets closed when we close the BucketStore. However, we currently don't close BucketStore
 		// on store-gateway termination so it never gets terminated.
 		goleak.IgnoreTopFunction("github.com/grafana/mimir/pkg/storegateway/indexheader.NewReaderPool.func1"),
+		// Similar to the one above, store-gateway BucketStore starts a goroutine in snapshotter
+		// that manages lazy-loaded snapshots. It stops the snapshotter on BucketStore close.
+		goleak.IgnoreTopFunction("github.com/grafana/mimir/pkg/storegateway/indexheader.(*Snapshotter).Start.func1"),
```

Comment:
Do you mean to have a proper lifecycle in `BucketStore` (and maybe `BucketStores`) so they managed the lifecycles of `ReaderPool` and `Snapshotter`? As you mentioned in https://github.com/grafana/mimir/pull/8281#discussion_r1638141640 (and what I'd bumped into when looked at it), this might be too large of a change for the expected gain.

I'm not opposed to doing that, though — refactoring can be fun 🕺

Reply:
yep that was my idea. Make `BucketStores` maintain the lifecycle of the `BucketStore` instances that it spawns - starting them when they come up, shutting them down when the store-gateway process shuts down. And then doing the appropriate things on these triggers (like starting the snapshotter; maybe even doing the initial sync as part of starting, which would make it slightly cleaner to use the lazy loaded snapshot -- for that we might bring back the snapshotter under ReaderPool). Maybe we can start simple and just managing the snapshotter and readerPool from BucketStore (and transitively from BucketStores)?



But I agree it's some refactoring. I'm happy to merge this PR so we catch the 2.13 and r294 releases on Monday, **as long as** we figure out how to make it work for tenants which are discovered. The rest is just code structure.